Editor feature to edit actor ID and owner.

drogoganor opened the pull request:
![actor edit](https://user-images.githubusercontent.com/1639681/44464808-189f5800-a65f-11e8-956e-4e5ff2625217.png)

A feature to select and edit an actor's ID and owner in the editor.

I need a bit of feedback on this one, especially on the following items:

1. I've added a `public EditorActorPreview SelectedActor` to EditorDefaultBrush, and HandleMouseInput() sets this if the user left-clicks an actor. I've also made the EditorViewportControllerWidget's defaultBrush public so that SelectedActor can be accessed and set by ActorEditLogic. I don't think this is a good design. I'm wondering if there's a nicer way of doing this.

2. I've made the EditorActorLayer's Add() public in order to facilitate removing and re-adding the actor in ActorEditLogic. I'm not sure if this is a good idea.

3. I've added an optional width constructor argument for SelectionBoxRenderable, which was hard-coded to 1. I found I needed to specify a wider width for the selection box because it was hard to see at just one pixel. But I'm not sure if this is appropriate or not. If you look closely at the image, you can see the left and right edges of the selection box are 2px extended.

4. In order to get the type of the actor and the nice tooltip name, I've had to add another field TypeSummary to EditorActorPreview in the same vein as Tooltip. Is this OK? Also I'd like to know if the formatting of this field in the image above is OK.

5. The color of the selection box. Right now you can see above it sort of blends in with the red buildings. Should this be another color? I tried hot pink, and that was certainly visible but...hot pink.

Cheers

reaperrr commented:
The formatting in the image looks good to me, although I'd maybe swap ID and Owner.
About the selection box color, have you tested neon green?

By the way, could you mabe add Health editing support as well?

drogoganor commented:
I've swapped ID and Owner, and also made the selection box green.

![limegreen](https://user-images.githubusercontent.com/1639681/44501754-dcf1a600-a6d1-11e8-8068-1067aeaf63ff.png)

It generally works pretty well except for tiberium as you can see, or snow:

![actor edit](https://user-images.githubusercontent.com/1639681/44501775-fbf03800-a6d1-11e8-86c3-7d0c4e9a2fc3.png)

And here's hot pink, which blends in with the d2k desert and probably cnc desert too, but generally stands out best:

![actor edit](https://user-images.githubusercontent.com/1639681/44501788-0c081780-a6d2-11e8-971d-9919ac2772d1.png)

MunWolf commented:
How about blue, not cyan though I imagine cyan will blend very easily with snow.

GraionDilach commented:
the infantry bounds is definitely related to #15148 and all it's followups.

pchote commented:
> the infantry bounds is definitely related to #15148 and all it's followups.

Kindof yes and kindof no - the ActorPreviews have never really needed a concept for interactable bounds, so this was never implemented - in its absence the only alternative is the size of the renderable which is indeed inflated for these infantry sprites.  This also affects the editor sidebar.

This all happened before the recent interaction bounds rework, so it should now be a lot easier to hook up the bounds defined in the actor rules.

KOYK commented:
Can you also add the ability to rotate the actor?

drogoganor commented:
Regarding blue selection box:

![actor edit](https://user-images.githubusercontent.com/1639681/44559401-2eab3680-a78d-11e8-8756-874f149bb96f.png)

It's OK too but blends in with grey a bit.

I'll give the alternating colors a go today.

Regarding obtaining the interaction bounds - I've tried getting InteractableInfo for the actor and trying to use the Bounds or DecorationBounds. But it seems to be almost always null. I'll reach out on IRC for some help.

Mailaender commented on OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs:
```diff
@@ -76,6 +79,8 @@ public EditorActorPreview(WorldRenderer worldRenderer, string id, ActorReference
 			Tooltip = (tooltip == null ? " < " + Info.Name + " >" : tooltip.Name) + "\n" + owner.Name + " (" + owner.Faction + ")"
 				+ "\nID: " + ID + "\nType: " + Info.Name;
 
+			TypeSummary = Info.Name + (tooltip != null ? " (" + tooltip.Name + ")" : string.Empty);
```
This adds a double `((` for things like the camera where there is already a `(` in the name.

Mailaender reviewed (dismissed):
Works as promised.

RobotCaleb commented:
Why not a 3 pixel wide selection caret?
black-white-black

drogoganor commented:
Updated to use a white selection box and flash the selected actor with the highlight palette every 500ms.

![image](https://user-images.githubusercontent.com/1639681/47156739-12f68380-d32b-11e8-8de5-986f69e17b0a.png)

pchote commented:
This is a high impact feature, so adding it to the milestone.

pchote commented on OpenRA.Mods.Common/Traits/World/EditorActorPreview.cs:
```diff
@@ -98,7 +107,9 @@ public void Tick()
-			return previews.SelectMany(p => p.Render(worldRenderer, CenterPosition));
+			var items = previews.SelectMany(p => p.Render(worldRenderer, CenterPosition))
+				.Select(y => !ShowBorder ? y : DateTime.Now.Millisecond < 500 ? y : y.WithPalette(worldRenderer.Palette("highlight")));
+			return !ShowBorder ? items : items.Concat(new IRenderable[] { SelectionBox });
```
In order to keep the visuals consistent with the rest of the game, would you mind changing this to:
```csharp
public IEnumerable<IRenderable> Render()
{
	var items = previews.SelectMany(p => p.Render(worldRenderer, CenterPosition));
	if (ShowBorder)
	{
		var highlight = worldRenderer.Palette("highlight");
		var overlay = items.Where(r => !r.IsDecoration)
			.Select(r => r.WithPalette(highlight));
		return items.Concat(overlay).Append(SelectionBox);
	}

	return items;
}
```
and keeping the selection box width as 1?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -0,0 +1,243 @@
+		readonly int originalHeight; // Height of the edit widget without the validation label showing.
+		readonly int validationHeight; // Height of the edit widget with the validation label showing.
+		readonly int editPanelPadding; // Padding between right edge of actor and the edit panel.
+		readonly TimeSpan scrollVisibleTimeout
```
We use `Game.RunTime` or `World.WorldTick` everywhere else in the game code for evaluating time.  It would be better if we could use one of these here instead of introducing a third method.

pchote reviewed (commented):
This looks great.  Just a couple of comments to start with:

drogoganor commented:
Here's what it looks like now. Rendering correctly!

![image](https://user-images.githubusercontent.com/1639681/47705326-f420b700-dc71-11e8-836d-2937b0ed1c2d.png)

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -0,0 +1,241 @@
+			actorDef.Key = newActorName;
+
+			// Remove existing editor actor
+			editorActorLayer.Remove(currentActor);
```
You should be able to `currentActor.ReplaceInit(new OwnerInit(selectedOwner.Name));` instead of serializing/removing/editing/readding the actor.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -0,0 +1,241 @@
+		{
+			if (currentActor != null)
+				editorActorLayer.Remove(currentActor);
+			Cancel();
```
General comment: try and be consistent about newlines after if statements (prefer having them).

pchote commented on OpenRA.Mods.Common/EditorBrushes/EditorDefaultBrush.cs:
```diff
@@ -27,12 +27,12 @@ public interface IEditorBrush : IDisposable
 	public sealed class EditorDefaultBrush : IEditorBrush
 	{
 		public readonly ActorInfo Actor;
-
```
Please keep this newline

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -0,0 +1,241 @@
+			{
+				// Selected actor is null, hide the border and edit panel.
+				if (currentActor != null)
+					currentActor.ShowBorder = false;
```
All this manual manipulation of `ShowBorder` / `Selected` is asking for trouble.  I suggest making `currentActor` a property, and defining a setter that does this automatically when the value is changed.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -0,0 +1,241 @@
+			// Check for duplicate actor ID
+			if (currentActor.ID.ToLowerInvariant() != newActorNameLower)
+			{
+				bool found = world.Map.ActorDefinitions.Any(x => x.Key.ToLowerInvariant() == newActorNameLower);
```
Style nit: use `var` instead of explicit types.

pchote commented:
From the screenshot above it looks like you are using the same margins for the TD panel as in the other mods.  Note that TD uses its own, smaller, margins because it does not use thick panel borders (see tooltips and the editor sidebar for examples).

drogoganor commented on OpenRA.Mods.Common/Widgets/Logic/Editor/ActorEditLogic.cs:
```diff
@@ -0,0 +1,241 @@
+				if (currentActor != actor)
+				{
+					lastScrollTime = 0; // Ensure visible
+					selectedOwner = editorActorLayer.Players.Players.Values.First(x => x.Name == actor.Owner.Name);
```
That would be much nicer!

drogoganor commented:
Here's the TD selection box made more compact with 3px margins like the actor selector:

![image](https://user-images.githubusercontent.com/1639681/47788887-bea5c780-dd5e-11e8-9a04-10c704bc22c0.png)

Phrohdoh commented:
Would it be possible to also change the actor type?

If not (because of a concern for, for example, size/footprint) then we can always do it in the future. :)

drogoganor commented:
Sort of like this as far as the Type name truncation goes?

![image](https://user-images.githubusercontent.com/1639681/48062144-de306a80-e20c-11e8-8083-595df0485717.png)

I'm not sure entirely what you mean by "it would be better all-round to use the same label as in the selection list."

drogoganor commented:
Updated to address issues. I have also made the EditorActorPreview's Tooltip field dynamic with a get property, since the ID and Owner can be set now. And updated the UI:

![image](https://user-images.githubusercontent.com/1639681/48066346-0d4cd900-e219-11e8-8445-4ace6e60f9d8.png)

drogoganor commented:
For your consideration, a mockup of how it would look with a bunch of other options:

![image](https://user-images.githubusercontent.com/1639681/48068015-70406f00-e21d-11e8-8764-848d20c24456.png)

pchote commented:
I have pushed over this with fixes for:
* Selection panel offset (was using the previews full width, instead of width / 2)
* A crash when actors don't define any tooltip traits

pchote commented:
I have a slight rework for the user interaction on 13928cbe120fe76c5cdf2ce918a636247f438882 / [pchote/map-editor-select](https://github.com/pchote/OpenRA/tree/map-editor-select).  The goal here was to streamline things looking forward to the (hopefully near) future where we expose additional parameters in the panel.

* Input is now applied immediately instead of losing them if you accidentally click out of the dialog.
* Cancel button has been replaced by Close, which behaves the same as clicking outside the dialog.
* Error labels are now shown immediately below the ID field.
* Widget layout slightly reorganized in preparation for dynamically populated parameters.

If I can get some :+1:s on these changes then we can merge them into this PR and then the whole feature into bleed.

reaperrr commented:
> If I can get some 👍s on these changes then we can merge them into this PR and then the whole feature into bleed.

👍 from me.

For those who don't want to/can't compile @pchote's changes, this is how the result looks like:
![pic01](https://user-images.githubusercontent.com/2857877/48870333-cba75b00-ede0-11e8-8231-0c46fea89094.png)

drogoganor commented:
Hi guys, I apologize for the unresponsiveness. Real life has made demands.

reaperrr commented:
> Hi guys, I apologize for the unresponsiveness. Real life has made demands.

No problem. Just to clarify, `For those who don't want to/can't compile @pchote's changes` was only directed at potential reviewers, not you.

drogoganor committed:
Add a basic actor properties panel to the editor.

pchote commented:
Rebased, squashed, and added a TODO to https://github.com/OpenRA/OpenRAModSDK/wiki/Update-notes:-release-20180923-to-next

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/264fa3419679e5055d90c0375a268bb9285a5d96
